### ISSUE TITLE
Blob generation ?

### ISSUE THREAD

**Issue by erdnaxe** [AndroidRoot/wheelie#1]
I'm trying to do nvflash on my tablet (ASUS SL101) to flash U-Boot. So I'm trying to get **wheelie working without the blob** which I didn't backup, and no working bootloader (long story, I was doing some kernel dev XD).

So far, I had 2 ideas : 
* **Method 1:** I can figure out the EMMC pinout (I need to buy an adaptor for 1.8V to sdcard), I will do this if the second method can't work to dump my blob and/or reflash my bootloader with my own U-Boot,
* **Method 2:** Maybe it's possible to modify Wheelie source code to **generate blob randomly** and try each one on the tablet, with a micro-controller automatically rebooting the tablet in APX mode.

The second method can maybe work but I need some help to modify wheelie. Reading the source code I understood that I need to work on these lines : 
```
if(!nvfblob_alloc_read(blob, NVFBLOB_TYPE_RCMVER, &buf, &bufsize)){
    printf("[-] Failed to read RCMVERS from blob file.\n");
    exit(0);
}

if(nvusb_send(nvdev, buf, bufsize) != 0) {
    printf("[-] RCM_Send communication failure.\n");
    exit(0);
}
```

So I only need to create a _nvfblob_header_t_ with a type of _NVFBLOB_TYPE_RCMVER_ ?
Then will I have to generate random data of this _length_ in _buf_ ?

Can you send a sample so I could figure out what I need to generate ?

I will stop trying random values when I get a response of _0x20001_ or _0x30001_ from the tablet.

If it's possible to crack blobs it will help lots of people (http://forum.xda-developers.com/showthread.php?t=2595381).

Regards,
Alexandre

**Comment by lilstevie** [comment]
Random blob generation will not work. The key used to encrypt the blob is AES128, there 2^128 possible keys. That corresponds to 3.4_10^38 combinations of keys. Assuming you struck it lucky and managed to brute force the key with just half the combinations that still works out to about 1.7_10^38 attempts. 

That is just the first problem you will face. The next one is that the USB port is shut down after every incorrectly encrypted message. To gain any hope of using a system with a microcontroller you need to be hooked in to hard disconnect the battery after every failure. The total time required for an attempt is 300 milliseconds for APX boot and verification of a blob. 

Using the inbuilt mechanism with the power button to restart takes 5 seconds which brings it to 5.3seconds per attempt which is simply way too long. Disconnecting the battery and starting again will only take about 50ms which brings your total attempt to 350ms. But even this isn't enough. 

This just reduces what would have taken 9.01_10^38 seconds (28.5 undecillion years) down to 5.95_10^37 seconds (1.8 undecillion years). This is only trying half of the possible combinations. We can't speed this up in any material way because the time it takes is the time for it to verify, and if we could remove the limitation from bootrom that shuts down the USB port (which would allow us to speed up the process by many orders of magnitude ) we would already have code execution and wouldn't need the blob. 

tl;dr; version. If you started this today, the big rip will have occurred long before you have cracked a single tablet.
